### PR TITLE
Put system information at the start of gamelog.txt

### LINKED ISSUE #1633: Suggestion: Move system info from title bar to gamelog.txt
I believe from discussion, the SDL, Lua version, Arch, and Direct 3D version are reported after CorsixTH in the game window's title for debugging purposes.

In reality when we want this info of this level for issues, we would ask for it in the form of the gamelog.txt
To improve the aesthetic of the window I would suggest we move this info into the gamelog. We could also probably stretch it out a bit to be a bit more informative about the data that matters.

The end result being the window title just becomes "CorisxTH" only; which is often the format other games present their game window.

### PR DESCRIPTION
**Pull request by tobylane**
*Previous feedback #1722*

**Describe what the proposed change does**
- Move the game log from world to app
- Log all available info to the game log, fix #1633, example

> Compiled with api_version: 2679, arch_64: true, os: macos, audio: false, 
Renderer: SDL, metal
Lua 5.4 run with game version: Trunk, savegame version: 152, api version: 2679

### CODE REVIEW

File: CorsixTH/Lua/entities/humanoid.lua [hunk 1]
```diff
@@ -494,10 +494,10 @@ local function Humanoid_startAction(self)
     end
 
     -- Tell the player what just happened.
-    self.world:gameLog("")
-    self.world:gameLog("Empty action queue!")
-    self.world:gameLog("Last action: " .. self.previous_action.name)
-    self.world:gameLog(debug.traceback())
+    self.world.app:gameLog("")
```
Comment (TheCycoONE) [inline CorsixTH/Lua/entities/humanoid.lua:497]: FWIW TheApp is a global, you don't need to go through world.

**Comment by tobylane** [comment]
Ready, unless you want that TheApp change.

**Comment by lewri** [comment]
Would personally recommend using ``TheApp`` but not up to me

**Comment by TheCycoONE** [comment]
Yeah, let's use TheApp for this. Nice and uniform. Then I can pretend it's a service for dependency injection and not a circular dependency.

**Comment by tobylane** [comment]
Done.

**Comment by TheCycoONE** [comment]
Play tested this a little. It appears to have lost the old gamelog data when I continue a game.
Here is a dump of the gamelog from my save from master:
```
Created game with savegame version 139.
Savegame version is 139 (Trunk), originally it was 139 (Trunk)
Savegame changed from 139 (Trunk) to 155 (Trunk). The save was created using 139 (Trunk).
Savegame version is 155 (Trunk), originally it was 139 (Trunk).
```

And here is the same game dumped from this PR.
```
Compiled with os: unix, arch_64: true, audio: true, api_version: 2679, 
Renderer: SDL, opengl
Lua 5.4 run with game version: Trunk, savegame version: 155, api version: 2679
Savegame version is 155 (Trunk), originally it was 139 (Trunk).
```

**Comment by tobylane** [comment]
I have a fix for that, but it's making me wonder if the system info should be put into every game log, and keep game log in World.

**Comment by TheCycoONE** [comment]
I was thinking the new system info should maybe just be added as a header when it gets dumped.  Open to other thoughts.

**Comment by TheCycoONE** [comment]
https://github.com/CorsixTH/CorsixTH/blob/edeba9fc767c4d011bb24ad48867694a1b09d6ab/CorsixTH/Lua/persistance.lua#L235 around here is where you would save/load a new top level object.

**Comment by tobylane** [comment]
Ready.

**Comment by TheCycoONE** [comment]
On my system this was the output:
```
Compiled with os: unix, renderer: SDL, opengl, arch_64: true, api_version: 2679, audio: true
Lua 5.4 run with game version: Trunk, savegame version: 155, api version: 2679
Created game with savegame version 155.
```

**Comment by lewri** [comment]
> arch_64: true

Why not just ``64-bit``? Though I would guess we also have ARM now... That should probably have its own identifier.

**Comment by tobylane** [comment]
That sounds good, I'll do that in another PR.

**Comment by TheCycoONE** [comment]
Can we avoid using the same separater that's used for renderer, or quote them.

**Comment by TheCycoONE** [comment]
also what's the purpose if having api_version listed twice.

**Comment by tobylane** [comment]
Replaced arch_64 with arch, outputs are x86, x64, ARM, ARM64 and Other. Any others worth including?

API version is from the C++ and Lua sides.

Removed the SDL part as it's unchanging, so this reports the SDL renderer.

> Compiled with arch: x64, audio: true, api_version: 2681, os: macos
SDL renderer: metal
Lua 5.4 run with game version: Trunk, savegame version: 158, api version: 2681

**Comment by TheCycoONE** [comment]
Ok, spiffy - thanks

**Comment by TheCycoONE** [comment]
Crashes on launch with MSVC:
```
An error has occurred in CorsixTH:
attempt to index a function value
stack traceback:
        [C]: in function 'TH.GetCompileOptions'
        E:/GitHub/CorsixTH/CorsixTH/Lua\app.lua:127: in method 'init'
        E:/GitHub/CorsixTH/CorsixTH/CorsixTH.lua:143: in main chunk
        [C]: in ?
```

File: CorsixTH/Src/th_lua.cpp [hunk 2]
```diff
@@ -185,15 +185,23 @@ int l_get_compile_options(lua_State* L) {
   lua_settop(L, 0);
   lua_newtable(L);
 
+#ifdef __x86_64__
```
Comment (TheCycoONE) [inline CorsixTH/Src/th_lua.cpp:188]: These re non-standard headers, they are not defined in MSVC

File: CorsixTH/Src/th_lua.cpp [hunk 3]
```diff
@@ -185,15 +185,23 @@ int l_get_compile_options(lua_State* L) {
...
 
-  lua_pushliteral(L, "SDL");
-  lua_setfield(L, -2, "renderer");
+#elif __arm__
+#ifdef CORSIX_TH_64BIT
+  lua_pushliteral(L, "ARM64");
+#else
+  lua_pushliteral(L, "ARM");
+#endif
+
```
Comment (TheCycoONE) [inline CorsixTH/Src/th_lua.cpp:201]: You are missing an else here, so Other isn't pushed when neither __x86_64__ nor __arm__ are defined.

**Review by TheCycoONE — CHANGES_REQUESTED** [review]
You'll need another approach to getting the compile host, I would suggest bringing it in from CMake instead of trying to use non-standard compiler headers.

https://cmake.org/cmake/help/latest/variable/CMAKE_HOST_SYSTEM_PROCESSOR.html as a start.

**Comment by tobylane** [comment]
Done.

> Compiled with api_version: 2681, audio: true, os: macos, arch: x86_64
SDL renderer: metal
Lua 5.4 run with game version: Trunk, savegame version: 158, api version: 2681

**Comment by TheCycoONE** [comment]
Looks good on my linux chromebook:
```
Compiled with os: unix, api_version: 2681, arch: aarch64, audio: true
SDL renderer: opengl
Lua 5.4 run with game version: Trunk, savegame version: 158, api version: 2681
```
Just want to test 32 and 64 bit windows builds before I merge.

**Comment by TheCycoONE** [comment]
Heh, I see we're seeing the non-deterministic order of pairs - our header orders don't match.

File: CorsixTH/Lua/app.lua [hunk 4]
```diff
@@ -1750,6 +1739,20 @@ function App:isAudioEnabled()
   return TH.GetCompileOptions().audio
 end
 
+function App:getSystemInfo()
+  local compile_opts = TH.GetCompileOptions()
+  local comp_details = ""
+  for key, value in pairs(compile_opts) do
```
Comment (TheCycoONE) [inline CorsixTH/Lua/app.lua:1745]: If we want a consistent order we'd need to build a list of keys and sort it (or hardcode it, but that's not as nice is it?)

Comment (Alberth289346) [reply CorsixTH/Lua/app.lua:1745]: Python2 also had non-deterministic dictionaries, and settled in Python3 for deterministic insertion order. Not sorted, but faster.

Simplest long term solution would be to change the lua table itself, but that may be out of scope, as you don't have a standard lua implementation any more in that case.

Sorting will work for tables with sortable keys, but likely not everything is fully ordered.

**Comment by tobylane** [comment]
The first line is sorted, the last line adjusted to match.

>Compiled with api_version: 2681, arch: x86_64, audio: true, os: macos
SDL renderer: metal
Lua 5.4 run with api version: 2681, game version: Trunk, savegame version: 158

File: CorsixTH/Lua/app.lua (hunk 4 above)
Comment (TheCycoONE) [reply CorsixTH/Lua/app.lua:1745]: I'm talking about a particular case where the UI (or gamelog header in this case) depended on the order of the pairs. That's a broader discussion - factorio for one modified lua's table to have a deterministic order, but that's likely out of scope for us. We can hack it with lists of keys where we need to (perhaps encapsulated)

**Comment by tobylane** [comment]
Ready.

**Comment by TheCycoONE** [comment]
Yes I know. I need to verify whether this works on Windows the way we package. There is a cmake issue about it, but I'm not sure if it applies the way I generate.

**Comment by lewri** [comment]
Here's mi windows portable from Appveyor's output if it helps:

```
Compiled with api_version: 2681, arch: AMD64, audio: true, os: windows
SDL renderer: direct3d
Lua 5.4 run with api version: 2681, game version: Trunk, savegame version: 158
```

**Comment by TheCycoONE** [comment]
This is the output from a Win32 (x86) build for Windows as we package it:

```
Compiled with api_version: 2681, arch: AMD64, audio: true, os: windows
SDL renderer: direct3d
Lua 5.4 run with api version: 2681, game version: Trunk, savegame version: 158
Created game with savegame version 158.
```

So yes, it seems to be broken there 😞

**Comment by TheCycoONE** [comment]
This is the upstream bug: https://gitlab.kitware.com/cmake/cmake/-/issues/15170

**Comment by tobylane** [comment]
We could keep arch_64 until that bug is fixed?

**Comment by TheCycoONE** [comment]
I tested this block and it seems to work (CMake documentation warns that it's internal - but...)
```
# Report system architecture in compile_opts.arch
if(MSVC)
  set(CORSIX_TH_ARCH ${CMAKE_CXX_COMPILER_ARCHITECTURE_ID})
else()
  set(CORSIX_TH_ARCH ${CMAKE_SYSTEM_PROCESSOR})
endif()
```

File: CorsixTH/CMakeLists.txt [hunk 5]
```diff
@@ -55,6 +55,13 @@ else()
   set(CORSIX_TH_OS "unix")
 endif()
 
+# Report system architecture in compile_opts.arch
+if(MSVC)
+  set(CORSIX_TH_ARCH ${CMAKE_CXX_COMPILER_ARCHITECTURE_ID})
```
Comment (TheCycoONE) [inline CorsixTH/CMakeLists.txt:60]: On discord I checked / pointed out that CMAKE_CXX_COMPILER_ARCHITECTURE_ID is CMake 3.10+ only. I don't mind bumping the cmake_minimum_version, or version guarding this condition.

**Pull request merged**